src/libraries/System.Net.Primitives/src/System/Net/IPNetwork.cs

```diff
@@ -0,0 +1,400 @@
+
+            if (PrefixLength == 0)
+            {
+                return true;
+            }
+
+            if (address.AddressFamily == AddressFamily.InterNetwork)
+            {
+                uint mask = uint.MaxValue << 32 - PrefixLength;
+
+                if (BitConverter.IsLittleEndian)
+                {
+                    mask = BinaryPrimitives.ReverseEndianness(mask);
+                }
+
+                return BaseAddress.PrivateAddress == (address.PrivateAddress & mask);
+            }
+            else
+            {
+                Unsafe.SkipInit(out UInt128 baseAddressValue);
+                Unsafe.SkipInit(out UInt128 otherAddressValue);
+
+                BaseAddress.TryWriteBytes(MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref baseAddressValue, 1)), out _);
+                address.TryWriteBytes(MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref otherAddressValue, 1)), out _);
+
+                UInt128 mask = UInt128.MaxValue << 128 - PrefixLength;
+                if (BitConverter.IsLittleEndian)
+                {
+                    mask = BinaryPrimitives.ReverseEndianness(mask);
+                }
+
+                return baseAddressValue == (otherAddressValue & mask);
+            }
+        }
+
+        /// <summary>
+        /// Converts a CIDR <see cref="string"/> to an <see cref="IPNetwork"/> instance.
+        /// </summary>
+        /// <param name="s">A <see cref="string"/> that defines an IP network in CIDR notation.</param>
+        /// <returns>An <see cref="IPNetwork"/> instance.</returns>
+        /// <exception cref="ArgumentNullException">The specified string is <see langword="null"/>.</exception>
+        /// <exception cref="FormatException"><paramref name="s"/> is not a valid CIDR network string, or the address contains non-zero bits after the network prefix.</exception>
+        public static IPNetwork Parse(string s)
+        {
+            ArgumentNullException.ThrowIfNull(s);
+            return Parse(s.AsSpan());
+        }
+
+        /// <summary>
+        /// Converts a CIDR character span to an <see cref="IPNetwork"/> instance.
+        /// </summary>
+        /// <param name="s">A character span that defines an IP network in CIDR notation.</param>
+        /// <returns>An <see cref="IPNetwork"/> instance.</returns>
+        /// <exception cref="FormatException"><paramref name="s"/> is not a valid CIDR network string, or the address contains non-zero bits after the network prefix.</exception>
+        public static IPNetwork Parse(ReadOnlySpan<char> s)
+        {
+            if (!TryParseCore(s, out IPAddress? address, out int prefixLength))
+            {
+                throw new FormatException(SR.net_bad_ip_network);
+            }
+
+            try
+            {
+                return new IPNetwork(address, prefixLength);
+            }
+            catch (Exception inner)
+            {
+                throw new FormatException(inner.Message, inner);
+            }
```

Comment:
```suggestion
            if (!TryParseCore(s, out IPAddress? address, out int prefixLength) ||
                !Validate(address, prefixLength, throwOnFailure: false))
            {
                throw new FormatException(SR.net_bad_ip_network);
            }
```
?  Then you can just change Validate to never throw.

Reply:
This way there wouldn't be an `InnerException` holding more information about the validation failure. (prefixLength out of range or "non-zero bits after the prefix".)

Reply:
If that's important, you can have Validate out an additional message to include in the FormatException.  We needn't throw an exception just to catch it and wrap it in something else.